astropy/cosmology/funcs.py

```diff
@@ -90,56 +125,130 @@ def z_at_value(func, fval, zmin=1e-8, zmax=1000, ztol=1e-8, maxfun=500):
+    if str(method).lower() == 'bounded':
+        opt['xatol'] = ztol
+        if bracket is not None:
+            warnings.warn(f"Option 'bracket' is ignored by method {method}.")
+            bracket = None
+    else:
+        opt['xtol'] = ztol
+
+    # fval falling inside the interval of bracketing function values does not guarantee
+    # it has a unique solution, but for Standard Cosmological quantities normally should
+    # (being monotonic or having a single extremum).
+    # In these cases keep solver from returning solutions outside of bracket.
+    fval_zmin, fval_zmax = func((zmin, zmax))
+    nobracket = False
     if np.sign(fval - fval_zmin) != np.sign(fval_zmax - fval):
-        warnings.warn("""\
-fval is not bracketed by func(zmin) and func(zmax). This means either
-there is no solution, or that there is more than one solution between
-zmin and zmax satisfying fval = func(z).""")
+        if bracket is None:
+            nobracket = True
+        else:
+            fval_brac = func(bracket)
+            if np.sign(fval - fval_brac[0]) != np.sign(fval_brac[-1] - fval):
+                nobracket = True
+            else:
+                zmin, zmax = bracket[0], bracket[-1]
+                fval_zmin, fval_zmax = fval_brac[[0, -1]]
+    if nobracket:
+        warnings.warn(f"fval is not bracketed by func(zmin)={fval_zmin} and func(zmax)="
+                      f"{fval_zmax}. This means either there is no solution, or that there is "
+                      "more than one solution between zmin and zmax satisfying fval = func(z).",
+                      AstropyUserWarning)
 
     if isinstance(fval_zmin, Quantity):
         val = fval.to_value(fval_zmin.unit)
-        f = lambda z: abs(func(z).value - val)
     else:
-        f = lambda z: abs(func(z) - fval)
-
-    zbest, resval, ierr, ncall = fminbound(f, zmin, zmax, maxfun=maxfun,
-                                           full_output=1, xtol=ztol)
-
-    if ierr != 0:
-        warnings.warn(f'Maximum number of function calls ({ncall}) reached')
-
-    if np.allclose(zbest, zmax):
-        raise CosmologyError("Best guess z is very close the upper z limit.\n"
+        val = fval
+
+    # 'Brent' and 'Golden' ignore `bounds`, force solution inside zlim
+    def f(z):
+        if z > zmax:
+            return 1.e300 * (1.0 + z - zmax)
+        elif z < zmin:
+            return 1.e300 * (1.0 + zmin - z)
+        elif isinstance(fval_zmin, Quantity):
+            return abs(func(z).value - val)
+        else:
+            return abs(func(z) - val)
+
+    res = minimize_scalar(f, method=method, bounds=(zmin, zmax), bracket=bracket, options=opt)
+
+    # Scipy docs state that `OptimizeResult` always has 'status' and 'message' attributes,
+    # but only `_minimize_scalar_bounded()` seems to have really implemented them.
+    if not res.success:
+        warnings.warn(f"Solver returned {res.get('status')}: {res.get('message', 'Unsuccessful')}\n"
```

Comment:
Perhaps a good use of a custom minimizer is one that tests this line! or maybe set the maximum number of evaluations to be super low and the starting point really far away?

Reply:
> @dhomeier, I think this is still an issue. func does not need to return a numpy array if given a non-scalar input.

Which one doesn't? Every single one I've checked start's right away with
```python
        if isiterable(z):
            z = np.asarray(z)
```
– I've probably not looked into all of them for all possible cosmology implementations, but this seemed quite clearly to define the standard API.

Reply:
I agree that it's a reasonable expectation that ``func(bracket)`` is amenable to numpy broadcasting and returns an ndarray. It's just that doesn't appear to be assumed in the old code.
In particular,
``fval_brac = func(bracket)`` can  be a list, not an ndarray. 
This passes all the new code except
``fval_zmin, fval_zmax = fval_brac[[0, -1]]``, which assumes an ndarray.
Changing it to 
``fval_zmin, fval_zmax = fval_brac[0],  fval_brac[-1]`` would also work on lists.

Reply:
> I agree that it's a reasonable expectation that `func(bracket)` is amenable to numpy broadcasting and returns an ndarray. It's just that doesn't appear to be assumed in the old code.

As I have yet to see an example of `func` that does not follow this convention (the tests are now covering all single-parameter functions of all cosmologies in `core`), this probably comes down to the question "what kind of cosmological functions is `z_at_value` supposed to handle". If this might include any hypothetical cosmology a user might provide, or even any generic function (the description "A function that takes redshift as input" is of course extremely generic), then it should be prepared to get lists, tuples or any kind of sequences returned. But my feeling is that a) this is opening a can or worms (might be talking about oscillating cosmological functions we have no guarantee how the solvers will behave for), and b) downstream may well expect to get an ndarray returned on any input that can instantiate one, so it would not hurt to detect non-conforming functions straight away.

Reply:
🆗 . Then this looks good! Right in time for v4.3 feature freeze! 🎉